src/main/java/org/openstreetmap/atlas/tags/filters/matcher/parsing/Lexer.java

```diff
@@ -0,0 +1,433 @@
+
+/**
+ * This class can transform an input {@link String} into a sequence of {@link Token}s recognizable
+ * by the {@link Parser}.
+ * 
+ * @author lcram
+ */
+public class Lexer
+{
+    /**
+     * @author lcram
+     */
+    private static class InputBuffer
+    {
+        static final int EOF = -1;
+
+        private final String string;
+        private int position;
+
+        InputBuffer(final String string)
+        {
+            this.string = string;
+            this.position = 0;
+        }
+
+        int consumeCharacter()
+        {
+            if (this.position >= this.string.length())
+            {
+                return EOF;
+            }
+            return this.string.charAt(this.position++);
+        }
+
+        int peek()
+        {
+            if (this.position >= this.string.length())
+            {
+                return EOF;
+            }
+            return this.string.charAt(this.position);
+        }
+
+        void unconsume()
+        {
+            if (this.position > 0)
+            {
+                this.position--;
+            }
+        }
+    }
+
+    /**
+     * @author lcram
+     */
+    private static class LexemeBuffer
+    {
+        private final List<Character> characters;
+
+        LexemeBuffer()
+        {
+            this.characters = new ArrayList<>();
+        }
+
+        @Override
+        public String toString()
+        {
+            final StringBuilder builder = new StringBuilder();
+            for (final Character character : this.characters)
+            {
+                builder.append(character);
+            }
+            return builder.toString();
+        }
+
+        void addCharacter(final char character)
+        {
+            this.characters.add(character);
+        }
+
+        void clear()
+        {
+            this.characters.clear();
+        }
+
+        LexemeBuffer stripLeading()
+        {
+            this.characters.remove(0);
+            return this;
+        }
+
+        LexemeBuffer stripTrailing()
+        {
+            this.characters.remove(this.characters.size() - 1);
+            return this;
+        }
+    }
+
+    public static String debugString(final List<Token> lexedTokens)
+    {
+        final StringBuilder builder = new StringBuilder();
+        for (final Token token : lexedTokens)
+        {
+            builder.append(token.toString());
+            builder.append(", ");
+        }
+        return builder.toString();
+    }
+
+    /**
+     * Lex a given input line.
+     *
+     * @param inputLine
+     *            the input line
+     * @return a {@link List} of the processed {@link Token}s
+     */
+    public List<Token> lex(final String inputLine) // NOSONAR
+    {
+        final List<Token> lexedTokens = new ArrayList<>();
+        final LexemeBuffer lexemeBuffer = new LexemeBuffer();
+        final InputBuffer inputBuffer = new InputBuffer(inputLine);
+        while (inputBuffer.peek() != InputBuffer.EOF)
```

Comment:
Wow, this method!

Reply:
Haha what about it? A lot of `if` statements? ;)